Added in interleave bits support

revans2 opened the pull request:
This adds in a new cuda kernel to interleave bits for zordering.

revans2 commented:
build

hyperbolic2346 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -0,0 +1,131 @@
+               cudf::size_type start_item_byte_index = input_row_number * base_size;
+               cudf::size_type read_byte_index = start_item_byte_index + (base_size - 1 - (le_read_byte_index - start_item_byte_index));
+
+               uint32_t byte_data = column.data<uint8_t>()[read_byte_index];
+               bool is_valid = (column.null_mask() ? cudf::bit_is_set(column.null_mask(), input_row_number + column.offset()): true);
+               if (!is_valid) {
+                 byte_data = 0;
+               }
```
```suggestion
               bool const is_valid = (column.null_mask() ? cudf::bit_is_set(column.null_mask(), input_row_number + column.offset()): true);
               uint32_t const byte_data = is_valid ? column.data<uint8_t>()[read_byte_index] : 0;
```

hyperbolic2346 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -0,0 +1,131 @@
+               if (!is_valid) {
+                 byte_data = 0;
+               }
+               uint32_t tmp = ((byte_data >> bit_offset) & 1) << ret_bit;
```
```suggestion
               uint32_t const tmp = ((byte_data >> bit_offset) & 1) << ret_bit;
```

hyperbolic2346 reviewed (changes requested):
Some comments on the c++, mostly just add const stuff.

abellina reviewed (commented):
First pass

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -0,0 +1,131 @@
+  const cudf::size_type max_bytes_allowed = std::numeric_limits<cudf::size_type>::max();
+
+  int64_t tmp_column_data_size = num_rows * base_size;
+  if (tmp_column_data_size > max_bytes_allowed) {
```
Yes it is needed because I am converting a column of INT to a column of binary. 2^31 INTs are allowed, but that results in 2^31 * 4 bytes, which is too large to fit.  Generally we should not allow this because we try to target only 2GiB columns at most, but if we have a batch that is too large for some reason we could run into issues.

revans2 commented on src/main/cpp/src/zorder.cu:
```diff
@@ -0,0 +1,131 @@
+             // Start with the highest bit for output
+             cudf::size_type ret_bit = 7;
+             uint8_t ret_byte = 0;
+             while (ret_bit >= 0) {
```
I copied and pasted from some scala code, and scala for loops are not as fast as while loops. I meant to fix it and never did. Thanks

revans2 commented on src/test/java/com/nvidia/spark/rapids/jni/ZOrderTest.java:
```diff
@@ -0,0 +1,385 @@
+      }
+      try (ColumnVector results = ZOrder.interleaveBits(numRows, cvInputs);
+           ColumnVector expectedCv = ColumnVector.fromLists(outputType, expected)) {
+        //System.err.println("EXPECTED: " + toHexString(expectedCv));
```
Yes I left them because of the toHexString, but if you want me to remove them all I will.

revans2 commented:
build

revans2 commented:
@abellina and @hyperbolic2346 I have addressed all of your concerns. If either of you have suggestions on how to break up the interleave bits so it is cleaner I am happy to implement them.

abellina reviewed (commented):
High level it makes sense, but I get a bit lost in the index math in the `for_each_n`. Personally, I think a few more comments would go a long way.

I really like the example in the test, it can be followed, and it sure is nice to see some of the same variable names in the kernel.

revans2 commented:
@abellina I think I have addressed the rest of your concerns. I have renamed a lot of the variables to hopefully explain better what they hold, but if you need more info, please let me know.

revans2 commented:
build

abellina reviewed (approved):
I am able to follow this a lot better. Thanks for the changes @revans2 , this looks good to me.
